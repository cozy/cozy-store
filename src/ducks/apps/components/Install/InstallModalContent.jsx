import React, { Component } from 'react'

import { ModalContent } from 'cozy-ui/react/Modal'
import Spinner from 'cozy-ui/react/Spinner'

import { translate } from 'cozy-ui/react/I18n'
import { SubTitle } from 'cozy-ui/react/Text'
import AnimatedModalHeader from 'ducks/components/AnimatedModalHeader'
import TransparencyLabel from '../TransparencyLabel'

export class InstallModalContent extends Component {
  render() {
    const { app, fetchError, isFetching, isCanceling, t } = this.props
    const isFirstLoading = isFetching && !isCanceling
    // this part must not be wrapped in a component
    // so we get the content using it as a function
    const animatedHeader = AnimatedModalHeader({
      app
    })

    return isFirstLoading ? (
      <ModalContent className="sto-modal-content">
        <div className="sto-install-loading">
          <Spinner size="xlarge" />
        </div>
      </ModalContent>
    ) : (
      <ModalContent className="sto-modal-content">
        {animatedHeader}
        <SubTitle className="sto-modal-title">
          {t('app_modal.install.title')}
        </SubTitle>
        <TransparencyLabel app={app} />
        {fetchError && (
          <p className="u-error">
            {t('app_modal.install.message.version_error', {
              message: fetchError.message
            })}
          </p>
        )}
      </ModalContent>
    )
  }
}

export default translate()(InstallModalContent)
