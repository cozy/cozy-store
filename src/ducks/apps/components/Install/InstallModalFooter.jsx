import React, { Component } from 'react'

import { ModalFooter } from 'cozy-ui/react/Modal'

import { translate } from 'cozy-ui/react/I18n'
import Icon from 'cozy-ui/react/Icon'

import cozySmileIcon from 'assets/icons/icon-cozy-smile.svg'

export class InstallModalFooter extends Component {
  installApp() {
    this.setState({ error: null })
    const { app, onError, channel } = this.props
    this.props.installApp(app.slug, app.type, channel).catch(error => {
      if (onError) return onError(error)
      throw error
    })
  }

  render() {
    const {
      fetchError,
      installError,
      isFetching,
      isInstalling,
      isCanceling,
      t
    } = this.props

    const isFirstLoading = isFetching && !isCanceling
    return (
      !isFirstLoading &&
      !fetchError && (
        <ModalFooter>
          {installError && (
            <p className="u-error">
              {t('app_modal.install.message.install_error', {
                message: installError.message
              })}
            </p>
          )}
          <div className="sto-install-controls">
            <button
              role="button"
              disabled={isInstalling || isCanceling}
              aria-busy={isInstalling}
              className="c-btn c-btn--regular c-btn--download"
              onClick={() => this.installApp()}
            >
              <Icon
                icon={cozySmileIcon}
                color="#FFFFFF"
                width="16px"
                height="16px"
                className="sto-app-install-button-icon"
              />
              <span>{t('app_modal.install.install')}</span>
            </button>
          </div>
        </ModalFooter>
      )
    )
  }
}

export default translate()(InstallModalFooter)
