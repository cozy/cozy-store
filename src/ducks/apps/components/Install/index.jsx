import React, { Component } from 'react'
import { withRouter } from 'react-router'

import Modal from 'cozy-ui/react/Modal'

import getChannel from 'lib/getChannelFromSource'

import InstallModalContent from './InstallModalContent'
import InstallModalFooter from './InstallModalFooter'

export class Install extends Component {
  constructor(props) {
    super(props)
    this.gotoParent = this.gotoParent.bind(this)
    this.state = {
      previousChannel: props.channel ? getChannel(props.app.source) : null,
      isCanceling: false
    }
    if (typeof props.fetchApp === 'function') props.fetchApp(props.channel)
  }

  componentWillReceiveProps(nextProps) {
    if (!nextProps.app) this.gotoParent()
  }

  async gotoParent() {
    const { app, parent, history, fetchApp } = this.props
    const { previousChannel } = this.state
    // fetch previous channel if channel switch canceled
    if (previousChannel && typeof fetchApp === 'function') {
      this.setState(() => ({ isCanceling: true }))
      await fetchApp(previousChannel)
      this.setState(() => ({ isCanceling: false }))
    }

    if (app && app.slug) {
      history.push(`${parent}/${app.slug}`)
    } else {
      history.push(parent)
    }
  }

  render() {
    const { app, installApp, isInstalling, channel, isAppFetching } = this.props
    const { isCanceling } = this.state
    if (!app) return null
    return (
      <div className="sto-modal--install">
        <Modal dismissAction={this.gotoParent} mobileFullscreen>
          <InstallModalContent
            app={app}
            isFetching={isAppFetching}
            isCanceling={isCanceling}
          />
          <InstallModalFooter
            app={app}
            installApp={installApp}
            isFetching={isAppFetching}
            channel={channel}
            isInstalling={isInstalling}
            isCanceling={isCanceling}
          />
        </Modal>
      </div>
    )
  }
}

export default withRouter(Install)
