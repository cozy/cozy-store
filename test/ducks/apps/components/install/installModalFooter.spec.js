'use strict'

/* eslint-env jest */

import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { tMock } from '../../../../jestLib/I18n'
import { InstallModalFooter } from 'ducks/apps/components/Install/InstallModalFooter'
import mockApp from '../../_mockPhotosRegistryVersion'

const mockInstallError = new Error('Install mock error')

Enzyme.configure({ adapter: new Adapter() })

const getProps = () => ({
  t: tMock,
  app: mockApp
})

describe('InstallModalFooter modal component', () => {
  it('should be rendered correctly', () => {
    const component = shallow(
      <InstallModalFooter {...getProps()} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should installApp on install button click', async () => {
    const props = getProps()
    props.channel = 'beta'
    props.installApp = jest.fn(() => Promise.resolve())
    const component = shallow(
      <InstallModalFooter {...props} installError={mockInstallError} />
    )
    const installButton = component.find('button')
    await installButton.simulate('click')
    expect(props.installApp.mock.calls.length).toBe(1)
    expect(props.installApp.mock.calls[0]).toMatchSnapshot()
  })

  it('should handle error when installApp', async () => {
    const props = getProps()
    props.channel = 'beta'
    const propsWithoutOnError = Object.assign({}, props)
    props.installApp = jest.fn(() => Promise.reject(mockInstallError))
    propsWithoutOnError.installApp = jest.fn(() =>
      Promise.reject(mockInstallError)
    )
    props.onError = jest.fn()
    const component = shallow(<InstallModalFooter {...props} />)
    const installButton = component.find('button')
    await installButton.simulate('click')
    expect(props.installApp.mock.calls.length).toBe(1)
    expect(props.installApp.mock.calls[0]).toMatchSnapshot()
    expect(props.onError.mock.calls.length).toBe(1)
    expect(props.onError.mock.calls[0][0]).toBe(mockInstallError)
    // should throw the error if onError not provided
    const component2 = shallow(<InstallModalFooter {...propsWithoutOnError} />)
    const installButton2 = component2.find('button')
    installButton2.simulate('click')
    expect(propsWithoutOnError.installApp.mock.calls.length).toBe(1)
    expect(propsWithoutOnError.installApp.mock.calls[0]).toMatchSnapshot()
  })

  it('should return nothing if isFetching', () => {
    const component = shallow(
      <InstallModalFooter {...getProps()} isFetching />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should have install button busy if isInstalling', () => {
    const component = shallow(
      <InstallModalFooter {...getProps()} isInstalling />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should have install button busy if isFetching but during a cancel (isCanceling)', () => {
    const component = shallow(
      <InstallModalFooter {...getProps()} isFetching isCanceling />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should display error if installError', () => {
    const component = shallow(
      <InstallModalFooter {...getProps()} installError={mockInstallError} />
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
