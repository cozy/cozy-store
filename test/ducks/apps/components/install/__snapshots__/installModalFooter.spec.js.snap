// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstallModalFooter modal component should be rendered correctly 1`] = `
<ModalFooter>
  <div
    className="sto-install-controls"
  >
    <button
      aria-busy={undefined}
      className="c-btn c-btn--regular c-btn--download"
      disabled={undefined}
      onClick={[Function]}
      role="button"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Accept and continue
      </span>
    </button>
  </div>
</ModalFooter>
`;

exports[`InstallModalFooter modal component should display error if installError 1`] = `
<ModalFooter>
  <p
    className="u-error"
  >
    Something went wrong when installing the application on your Cozy. Reason: Install mock error
  </p>
  <div
    className="sto-install-controls"
  >
    <button
      aria-busy={undefined}
      className="c-btn c-btn--regular c-btn--download"
      disabled={undefined}
      onClick={[Function]}
      role="button"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Accept and continue
      </span>
    </button>
  </div>
</ModalFooter>
`;

exports[`InstallModalFooter modal component should handle error when installApp 1`] = `
Array [
  "photos",
  "webapp",
  "beta",
]
`;

exports[`InstallModalFooter modal component should handle error when installApp 2`] = `
Array [
  "photos",
  "webapp",
  "beta",
]
`;

exports[`InstallModalFooter modal component should have install button busy if isFetching but during a cancel (isCanceling) 1`] = `
<ModalFooter>
  <div
    className="sto-install-controls"
  >
    <button
      aria-busy={undefined}
      className="c-btn c-btn--regular c-btn--download"
      disabled={true}
      onClick={[Function]}
      role="button"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Accept and continue
      </span>
    </button>
  </div>
</ModalFooter>
`;

exports[`InstallModalFooter modal component should have install button busy if isInstalling 1`] = `
<ModalFooter>
  <div
    className="sto-install-controls"
  >
    <button
      aria-busy={true}
      className="c-btn c-btn--regular c-btn--download"
      disabled={true}
      onClick={[Function]}
      role="button"
    >
      <Icon
        className="sto-app-install-button-icon"
        color="#FFFFFF"
        height="16px"
        icon="test-file-stub"
        spin={false}
        width="16px"
      />
      <span>
        Accept and continue
      </span>
    </button>
  </div>
</ModalFooter>
`;

exports[`InstallModalFooter modal component should installApp on install button click 1`] = `
Array [
  "photos",
  "webapp",
  "beta",
]
`;

exports[`InstallModalFooter modal component should return nothing if isFetching 1`] = `null`;
